sklearn/metrics/_pairwise_distances_reduction/_argkminlabels.pyx

```diff
@@ -0,0 +1,203 @@
+
+        # Heap to be returned
+        self.argkmin_labels = np.full(self.n_samples_X, -1, dtype=ITYPE)
+
+        # Map from set of unique labels to their indices in `label_counts`
+        self.labels_to_index = {label:idx for idx, label in enumerate(self.unique_labels)}
+        # Buffer used in building a histogram for one-pass weighted mode
+        self.label_counts = np.zeros((self.n_samples_X,  len(self.unique_labels)), dtype=DTYPE)
+
+    def _finalize_results(self, bint return_distance=False):
+        if return_distance:
+            # We need to recompute distances because we relied on
+            # surrogate distances for the reduction.
+            self.compute_exact_distances()
+
+            # Values are returned identically to the way `KNeighborsMixin.kneighbors`
+            # returns values. This is counter-intuitive but this allows not using
+            # complex adaptations where `PairwiseDistancesArgKmin64.compute` is called.
+            return np.asarray(self.argkmin_distances), np.asarray(self.argkmin_indices), np.asarray(self.argkmin_labels)
+
+        return np.asarray(self.argkmin_indices), np.asarray(self.argkmin_labels)
+
+    cdef void _parallel_on_X_prange_iter_finalize(
+        self,
+        ITYPE_t thread_num,
+        ITYPE_t X_start,
+        ITYPE_t X_end,
+    ) nogil:
+        cdef:
+            ITYPE_t idx, jdx, y_idx, label_index, sample_index
+            DTYPE_t max_weight
+            ITYPE_t label, max_label
+            DTYPE_t val, weight
+        # Sorting the main heaps portion associated to `X[X_start:X_end]`
+        # in ascending order w.r.t the distances.
+        for idx in range(X_end - X_start):
+            simultaneous_sort(
+                self.heaps_r_distances_chunks[thread_num] + idx * self.k,
+                self.heaps_indices_chunks[thread_num] + idx * self.k,
+                self.k
+            )
+            # One-pass top-one weighted mode
+            sample_index = thread_num * self.X_n_samples_chunk + idx
+            max_weight = -1
+            for jdx in range(self.k): # Iterate through k-nearest neighbors
+                if self.weight_type == Weight.uniform:
+                    weight = 1.
+                elif self.weight_type == Weight.distance:
+                    weight = 1. / self.heaps_r_distances_chunks[thread_num][idx*self.k+jdx]
+
+                y_idx = self.heaps_indices_chunks[thread_num][idx*self.k+jdx]
+                label = self.labels[y_idx]
+                label_index = self.labels_to_index[label]
+                self.label_counts[sample_index][label_index] += weight
+                val = self.label_counts[sample_index][label_index]
+                if max_weight < val or (max_weight == val and label < max_label):
+                    max_label = label
+                    max_weight = val
+            self.argkmin_labels[thread_num*self.X_n_samples_chunk + idx] = max_label
+
+    cdef void _parallel_on_Y_finalize(
```

Comment:
Suggestions for `_parallel_on_X_prange_iter_finalize` also apply here.

I think `idx` can be renamed to `samples_index` for consistency.